### PR TITLE
56117 - Implement basic receipts in the Italian Middleware

### PR DESCRIPTION
**Pull request by forsthug**
- [x] You've read the [Contributor Guide](https://github.com/fiskaltrust/.github/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/fiskaltrust/.github/blob/main/CODE_OF_CONDUCT.md).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

{Summary of the changes}

## Description

{Detail}

Fixes #{issue number}

### CODE REVIEW

File: queue/src/fiskaltrust.Middleware.Localization.QueueIT/RequestCommands/InitialOperationReceiptCommand.cs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        private readonly ftQueueIT _queueIt;
+
+        public InitialOperationReceiptCommand(ILogger<InitialOperationReceiptCommand> logger, IConfigurationRepository configurationRepository, SignatureItemFactoryIT signatureItemFactoryIT, ftQueueIT queueIt) : base(logger)
+        {
+            _configurationRepository = configurationRepository;
+            _signatureItemFactoryIT = signatureItemFactoryIT;
+            _queueIt = queueIt;
+        }
+
+        protected override async Task<(ftActionJournal, SignaturItem)> InitializeSCUAsync(ftQueue queue, ReceiptRequest request, ftQueueItem queueItem)
```
Comment (TSchmiedlechner) [inline queue/src/fiskaltrust.Middleware.Localization.QueueIT/RequestCommands/InitialOperationReceiptCommand.cs:28]: We also need to call the base method here, right?

File: queue/src/fiskaltrust.Middleware.Localization.QueueIT/RequestCommands/OutOfOperationReceiptCommand.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+        public OutOfOperationReceiptCommand(SignatureItemFactoryIT signatureItemFactoryIT, ftQueueIT queueIt)
+        {
+            _signatureItemFactoryIT = signatureItemFactoryIT;
+            _queueIt = queueIt;
+        }
+
+        public override long CountryBaseState => Constants.Cases.BASE_STATE;
+
+        protected override Task<(ftActionJournal, SignaturItem)> DeactivateSCUAsync(ftQueue queue, ReceiptRequest request, ftQueueItem queueItem)
```
Comment (TSchmiedlechner) [inline queue/src/fiskaltrust.Middleware.Localization.QueueIT/RequestCommands/OutOfOperationReceiptCommand.cs:24]: We also need to call the base method here, right?

File: queue/src/fiskaltrust.Middleware.Localization.QueueIT/RequestCommands/PosReceiptCommand.cs [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        {
+            _client = itIsscdProvider.Instance;
+            _signatureItemFactoryIT = signatureItemFactoryIT;
+        }
+        public override async Task<RequestCommandResponse> ExecuteAsync(ftQueue queue, ReceiptRequest request, ftQueueItem queueItem)
+        {
+            var receiptResponse = CreateReceiptResponse(queue, request, queueItem, CountryBaseState);
+
+            FiscalReceiptResponse response;
+            if (!request.IsVoid())
```
Comment (TSchmiedlechner) [inline queue/src/fiskaltrust.Middleware.Localization.QueueIT/RequestCommands/PosReceiptCommand.cs:31]: Could we reverse this if? :D

File: queue/src/fiskaltrust.Middleware.Localization.QueueIT/RequestCommands/PosReceiptCommand.cs [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+            else
+            {
+                var fiscalReceiptRefund = CreateRefund(request);
+                response = await _client.FiscalReceiptRefundAsync(fiscalReceiptRefund).ConfigureAwait(false);
+            }
+            receiptResponse.ftSignatures = _signatureItemFactoryIT.CreatePosReceiptSignatures(response);
+
+            if (!response.Success)
+            {
+                throw new Exception(response.ErrorInfo);
```
Comment (TSchmiedlechner) [inline queue/src/fiskaltrust.Middleware.Localization.QueueIT/RequestCommands/PosReceiptCommand.cs:45]: Let's maybe add some more information here, or (better) use a dedicated exception class for this.

File: queue/src/fiskaltrust.Middleware.Localization.QueueIT/RequestCommands/PosReceiptCommand.cs [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+                ReceiptResponse = receiptResponse,
+                ActionJournals = new List<ftActionJournal>()
+            };
+        }
+
+        private static FiscalReceiptInvoice CreateInvoice(ReceiptRequest request)
+        {
+            var fiscalReceiptRequest = new FiscalReceiptInvoice()
+            {
+                //TODO Barcode = "0123456789"
```
Comment (TSchmiedlechner) [inline queue/src/fiskaltrust.Middleware.Localization.QueueIT/RequestCommands/PosReceiptCommand.cs:59]: Barcode can be read from `ChargeItem.ProductBarcode`

File: queue/src/fiskaltrust.Middleware.Localization.QueueIT/RequestCommands/PosReceiptCommand.cs [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+                }).ToList()
+            };
+            return fiscalReceiptRequest;
+        }
+
+        private static FiscalReceiptRefund CreateRefund(ReceiptRequest request)
+        {
+            var fiscalReceiptRequest = new FiscalReceiptRefund()
+            {
+                //TODO Barcode = "0123456789"
```
Comment (TSchmiedlechner) [inline queue/src/fiskaltrust.Middleware.Localization.QueueIT/RequestCommands/PosReceiptCommand.cs:84]: Barcode can be read from `ChargeItem.ProductBarcode`

File: queue/src/fiskaltrust.Middleware.Localization.QueueIT/RequestCommands/PosReceiptCommand.cs [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+                //TODO Barcode = "0123456789" 
+                //TODO DisplayText = "Message on customer display",
+                Refunds = request.cbChargeItems?.Select(p => new Refund
+                {
+                    Description = p.Description,
+                    Quantity = Math.Abs(p.Quantity),
+                    UnitPrice = p.UnitPrice ?? 0,
+                    Amount = Math.Abs(p.Amount),
+                    VatGroup = p.GetVatGroup(),
+                    OperationType = p.GetRefundOperationType()
```
Comment (TSchmiedlechner) [inline queue/src/fiskaltrust.Middleware.Localization.QueueIT/RequestCommands/PosReceiptCommand.cs:93]: This method can return `null`, is this okay here? If not, what should we use as the default value?

File: queue/src/fiskaltrust.Middleware.Localization.QueueIT/RequestCommands/RequestCommandIT.cs [hunk 8]
```diff
@@ -0,0 +1,9 @@
+﻿using fiskaltrust.Middleware.Contracts.RequestCommands;
+
+namespace fiskaltrust.Middleware.Localization.QueueIT.RequestCommands
+{
+    public abstract class RequestCommandIT : RequestCommand
```
Comment (TSchmiedlechner) [inline queue/src/fiskaltrust.Middleware.Localization.QueueIT/RequestCommands/RequestCommandIT.cs:5]: Are we using this class somewhere?

File: queue/src/fiskaltrust.Middleware.Localization.QueueIT/SignProcessorIT.cs [hunk 9]
```diff
@@ -3,25 +3,47 @@
...
+            _configurationRepository = configurationRepository;
         }
 
-        public Task<(ReceiptResponse receiptResponse, List<ftActionJournal> actionJournals)> ProcessAsync(ReceiptRequest request, ftQueue queue, ftQueueItem queueItem)
+        public async Task<(ReceiptResponse receiptResponse, List<ftActionJournal> actionJournals)> ProcessAsync(ReceiptRequest request, ftQueue queue, ftQueueItem queueItem)
         {
-            throw new NotImplementedException();
+            var queueIT = await _configurationRepository.GetQueueITAsync(queue.ftQueueId).ConfigureAwait(false);
+
+            queueIT ??= new ftQueueIT() { ftQueueITId = queue.ftQueueId};
```
Comment (TSchmiedlechner) [inline queue/src/fiskaltrust.Middleware.Localization.QueueIT/SignProcessorIT.cs:42]: We should remove this, configuration objects should never be created at runtime. These are configured in the Portal and must be written to the DB at bootstrap time :)

File: queue/src/fiskaltrust.Middleware.Storage.MySQL/Migrations/008_QueueITuES.sqlite3 [hunk 10]
```diff
@@ -0,0 +1,41 @@
+﻿CREATE TABLE IF NOT EXISTS ftQueueES
```
Comment (TSchmiedlechner) [inline queue/src/fiskaltrust.Middleware.Storage.MySQL/Migrations/008_QueueITuES.sqlite3:1]: Maybe we should leave all the Spanish stuff out for now? We don't know yet if these models will look like this, I think?

File: queue/src/fiskaltrust.Middleware.Localization.QueueIT/Extensions/ChargeItemExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+{
+    public static class ChargeItemExtensions
+    {
+        public static OperationType? GetRefundOperationType(this ChargeItem chargeItem)
+        {
+            return (chargeItem.ftChargeItemCase & 0x0000_0000_000F_0000) switch
+            {
+                0x10000 => OperationType.Acconto,
+                0x20000 => OperationType.FreeOfCharge,
+                0x30000 => OperationType.SingleUseVoucher,
```
[the remark this answers is not in the document]
Comment (forsthug) [reply queue/src/fiskaltrust.Middleware.Localization.QueueIT/Extensions/ChargeItemExtensions.cs:15]: (chargeItem.ftChargeItemCase & 0x0000_0000_000F_0000) gives x_0000

File: queue/src/fiskaltrust.Middleware.Localization.QueueIT/RequestCommands/InitialOperationReceiptCommand.cs (hunk 1 above)
Comment (forsthug) [reply queue/src/fiskaltrust.Middleware.Localization.QueueIT/RequestCommands/InitialOperationReceiptCommand.cs:28]: the base is calling this

**Review by volllly — APPROVED** [review]

**Pull request merged**